### ISSUE TITLE
warn  - You should not access 'res' after getServerSideProps resolves.

### ISSUE THREAD

**Issue by hacknlove** [auth0/nextjs-auth0#524]
### Description

next12 warns

> warn  - You should not access 'res' after getServerSideProps resolves.
> Read more: https://nextjs.org/docs/messages/gssp-no-mutating-res

### Reproduction
Using withPageAuthRequired or getSession is enough to get the warning

### Environment

nextjs@12

**Comment by viperfx** [comment]
Getting similar warning with nextjs 12 as well. But login so far seems to functionally work.

**Comment by Seijinx** [comment]
I think it may break in a future update. Since Next 12, HOC should be
replaced by the middleware approach proposed here:

https://nextjs.org/docs/middleware

**Comment by adikari** [comment]
Any updates on this?

**Comment by adamjmcgrath** [comment]
Hi @adikari - thanks for checking. No update, we'll update this thread when we do

**Comment by davecarlson** [comment]
I take it a solution for this would be to move to using the next middleware instead ?

**Comment by robotastronaut** [comment]
So, I ran into this myself and decided to look into it, because it struck me as a notable change in the bump to v12. First of all, this error is *only* thrown in a [non-production environment](https://github.com/vercel/next.js/blob/canary/packages/next/server/render.tsx#L923): 

```JavaScript
if (process.env.NODE_ENV !== 'production') {
  resOrProxy = new Proxy<ServerResponse>(res, {
    get: function (obj, prop, receiver) {
      if (!canAccessRes) {
        const message =
          `You should not access 'res' after getServerSideProps resolves.` +
          `\nRead more: https://nextjs.org/docs/messages/gssp-no-mutating-res`

        if (deferredContent) {
          throw new Error(message)
        } else {
          warn(message)
        }
      }
      return Reflect.get(obj, prop, receiver)
    },
  })
}
```

You'll note here that what it's doing is creating a Proxy of the original `res` object in a non-prod environment specifically to monitor this scenario -- everywhere else in this function, the response is referenced by `resOrProxy`. All that happens in the scenario in this issue is a warning being displayed. An error will be thrown if `deferredContent` is truthy, which happens when [`data.props` is a `Promise`](https://github.com/vercel/next.js/blob/canary/packages/next/server/render.tsx#L975):

```JavaScript
if ((data as any).props instanceof Promise) {
  deferredContent = true
}
```

If that's considered the scenario worth throwing an exception for, then I think we can keep using this as-is. The library that is used to manipulate headers on the response is `on-headers`, which adds listeners to the underlying response object to respond to headers finally being emitted (and thus available for manipulation). This event fires after Next has wrapped up doing what it needs to do and `res.writeHead` is called either explicitly by Next (I think it does this on redirects and a few other places), or implicitly by the underlying `http_server` on `res.write` or `res.end`.

Setting the auth cookie is one of the last things happening to the response on the way out the door, and I can't see any scenarios in which it would actually cause a conflict with anything Next is doing. This safeguard appears to be in place to protect the response from being manipulated in conflicting ways if `getServerSideProps` somehow manages to return an object where `props` is a promise, but it's being awaited at every point. I feel like you'd need to be deliberately breaking something or being impressively careless to trigger that exception.

I suppose the cookie could be set at some point in the auth required factories directly rather than using `on-header` but that would eliminate most of the convenience of this library and risk headers being overwritten downstream.

I tried really hard to get this to break in some meaningful way, but I couldn't. For my use case, anyways, this is very much a "nothing to see here, move along" scenario until Next turns that `warn` into an exception. Certainly is noisy, though.

Would love to hear of any scenario in which this actually causes problems.

**Comment by ioandev** [comment]
To temporarily fix this, go in your `next.config.js` and add this before the `module.exports` line:
```js
let consolewarn = console.warn
console.warn = (...args) => {
    if (args.length > 1 && args[1].startsWith("You should not access 'res' after getServerSideProps")) {
        // ignore message until this is fixed: https://github.com/auth0/nextjs-auth0/issues/524
    } else {
        consolewarn(...args)
    }
}
```

**Comment by sohamsshah** [comment]
> To temporarily fix this, go in your `next.config.js` and add this before the `module.exports` line:
> 
> ```js
> let consolewarn = console.warn
> console.warn = (...args) => {
>     if (args.length > 1 && args[1].startsWith("You should not access 'res' after getServerSideProps")) {
>         // ignore message until this is fixed: https://github.com/auth0/nextjs-auth0/issues/524
>     } else {
>         consolewarn(...args)
>     }
> }
> ```

Elite chad

**Comment by JevAr00** [comment]
Possible solution.

Next.js (with js files) use a different way to handle the getServerSideProps.
The example that developers give to us use jsx files and I think that they are using react.js (I'm learning to use react and next so... 50% sure about this opinion)
I'm developing a small university project and I found this solution.

I didn't get more warnings.
![image](https://user-images.githubusercontent.com/89232748/157387191-2dc60c51-bfce-41ab-8a8d-55901af24698.png)

**Comment by eatrocks** [comment]
@JevAr00, I tried your possible solution of calling `withPageAuthRequired()` within `getServerSideProps`, but the page is not protected; authentication is not required.

**Comment by JevAr00** [comment]
> @JevAr00, I tried your possible solution of calling `withPageAuthRequired()` within `getServerSideProps`, but the page is not protected; authentication is not required.

Yeap, I was trying to solve that, if I found a solution I'm going to post it.

**Comment by jmunckhof** [comment]
Same issue here

**Comment by wicka-aus** [comment]
Getting same warning with withPageAuthRequired

**Comment by maxpaj** [comment]
Same issue here!

**Comment by Paola3stefania** [comment]
Hello !
After reading and researching, the best implementation that I found to fix this is encapsulating the page function with the auth0 hook:

export default **withPageAuthRequired**(function YourPage({ user }) {
	return(
	<div> Hi im fixed </div>)
	}

Hope it helps !
Happy Coding :)

**Comment by adamjmcgrath** [comment]
Thanks for your answers and workarounds on this thread. We'll be looking at some solutions around session lifecycle to remove this warning in a next major release. I don't have a timeline on when we may work on this, but you should use one of the workarounds in this thread or ignore the warning for now.

**Comment by remram44** [comment]
Is there another issue we can subscribe to, to be notified of your progress? A bit surprised to see this issue closed while the problem remained, is this no longer where issues are tracked?

**Comment by adamjmcgrath** [comment]
Hi @remram44 - thanks for your feedback

Sure, we can reopen this until a better way to track the next major comes along

**Comment by adamjmcgrath** [comment]
@zackdotcomputer - let me have another think about it this week. We may be able to implement a temporary measure before we do a breaking change.

**Comment by tangye1234** [comment]
When will it be resolved? It is too bothering. I have used nextjs for nearly 4 years, lot of our utils are using on-headers / on-finished, which dose no harm to next.

**Comment by adamjmcgrath** [comment]
This should be fixed in [1.19.0](https://github.com/auth0/nextjs-auth0/releases/tag/v1.9.0)

Have a look at https://github.com/auth0/nextjs-auth0/blob/main/FAQ.md#3-im-getting-the-warningerror-you-should-not-access-res-after-getserversideprops-resolves if you're still seeing warnings because you're using `getSession`/`getAccessToken` without `withPageAuthRequired`.

**Comment by zackdotcomputer** [comment]
Hmm even after #664 I'm having issues. Minor issue is that the types on the handler are getting erased, filed in #675. A more serious issue, though, is that when I wrap my handler with this wrapper then my page stops rendering with the following error:
```
TypeError: "baseURL" is required
Uncaught     at Object.get (file:///Users/zack/code/addition/web/node_modules/@auth0/nextjs-auth0/dist/auth0-session/get-config.js:147:15)
    at Object.getConfig (file:///Users/zack/code/addition/web/node_modules/@auth0/nextjs-auth0/dist/config.js:66:38)
    at Object._initAuth (file:///Users/zack/code/addition/web/node_modules/@auth0/nextjs-auth0/dist/index.js:22:23)
    at getInstance (file:///Users/zack/code/addition/web/node_modules/@auth0/nextjs-auth0/dist/index.js:18:24)
    at getSessionCache (file:///Users/zack/code/addition/web/node_modules/@auth0/nextjs-auth0/dist/index.js:62:44)
    at wrappedGetServerSideProps (file:///Users/zack/code/addition/web/node_modules/@auth0/nextjs-auth0/dist/helpers/get-server-side-props-wrapper.js:14:32)
    at Object.renderToHTML (file:///Users/zack/code/addition/web/node_modules/next/dist/server/render.js:615:26)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async doRender (file:///Users/zack/code/addition/web/node_modules/next/dist/server/base-server.js:901:38)
```

Is something amiss with the function or is this a PEBKAC with how i'm using it?

**Comment by adamjmcgrath** [comment]
Hey @zackdotcomputer 

Thanks for raising https://github.com/auth0/nextjs-auth0/issues/675

> Is something amiss with the function or is this a PEBKAC with how i'm using it?

Not sure, how are you using it?

**Comment by zackdotcomputer** [comment]
@adamjmcgrath We're using RBAC passed via the token's scope, so we have a custom analog to `withPageAuthenticationRequired` that can check for roles and 404 if the user can't view the page in question and which passes the roles in to the wrapped server-side props function.

I uploaded a gist with a simplified version, showing the usage of `getServerSidePropsWrapper`: https://gist.github.com/zackdotcomputer/9f2d57d0ebc3cced3c52fca83fed1d1c

**Comment by adamjmcgrath** [comment]
> EDIT: nope - it was PEBCAK. We were instantiating the Auth0 client manually as described in the docs. This meant that we had to use our custom client's getServerSidePropsWrapper rather than the one exported from the package itself.

Thanks for the update 👍 (I need to figure out how to add a warning for those cases)

**Comment by fabiolnm** [comment]
Is this warning still expected?
```
  "dependencies": {
    "@auth0/nextjs-auth0": {
      "version": "1.9.1",
```

Code:
```
import { withPageAuthRequired, getSession } from '@auth0/nextjs-auth0'

export async function getServerSideProps({ req, res }) {
  const session = getSession(req, res)
  const { name } = session.user
  console.log('name', name)
  return {
    props: { name }, 
  }
}

export default withPageAuthRequired(({ name }) => <>Hello, {name}!</>)
```

Output:
```
name Fabio Miranda
warn  - You should not access 'res' after getServerSideProps resolves.
Read more: https://nextjs.org/docs/messages/gssp-no-mutating-res
```

**Comment by adamjmcgrath** [comment]
Hi @fabiolnm - have a look at https://github.com/auth0/nextjs-auth0/blob/main/FAQ.md#3-im-getting-the-warningerror-you-should-not-access-res-after-getserversideprops-resolves

**Comment by lukben2000** [comment]
The link is not available anymore. For anyone that might still need this, I found the historic page with additional details using wayback machine: https://web.archive.org/web/20221011110725/https://github.com/auth0/nextjs-auth0/blob/main/FAQ.md#3-im-getting-the-warningerror-you-should-not-access-res-after-getserversideprops-resolves

The solution that worked for me was to wrap my getServerSideProps in `getServerSidePropsWrapper()`
